sycl/include/CL/sycl/handler.hpp

```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
+  /// a kernel name for it.
+  ///
+  /// TODO: currently it calls only those versions of kernels that can handle
+  /// custom types and operations. Some of types and operations may use faster
+  /// implementations that use intel::reduce() and/or sycl::atomic.fetch_<op>()
+  /// functions and thus provide much better performance. Those variants exist,
+  /// are fully functional. They just wait for their time for code-review.
+  /// TODO: Need to handle more than 1 reduction in parallel_for().
+  /// TODO: Support HOST. The kernels called by this parallel_for() may use
+  /// some functionality that is not yet supported on HOST such as:
+  /// barrier(), and intel::reduce() that also may be used in more
+  /// optimized implementations waiting for their turn of code-review.
+  template <typename KernelName = detail::auto_name, typename KernelType,
+            int Dims, typename Reduction>
+  void parallel_for(nd_range<Dims> Range, Reduction &Redu,
+                    KernelType KernelFunc) {
+    size_t NWorkGroups = Range.get_group_range().size();
+
+    // This parallel_for() is lowered to the following sequence:
+    // 1) Call a kernel that a) call user's lambda function and b) performs
+    //    one iteration of reduction, storing the partial reductions/sums
+    //    to either a newly created global buffer or to user's reduction
+    //    accessor. So, if the original 'Range' has totally
+    //    N1 elements and work-group size is W, then after the first iteration
+    //    there will be N2 partial sums where N2 = N1 / W.
+    //    If (N2 == 1) then the partial sum is written to user's accessor.
+    //    Otherwise, a new global buffer is created and partial sums are written
+    //    to it.
+    // 2) Call an aux kernel (if necessary, i.e. if N2 > 1) as many times as
+    //    necessary to reduce all partial sums into one final sum.
+
+    // 1. Call the kernel that includes user's lambda function.
+    // If this kernel is going to be now last one, i.e. it does not write
+    // to user's accessor, then detach user's accessor from this kernel
+    // to make the dependencies between accessors and kernels more clean and
+    // correct.
+    if (NWorkGroups > 1)
+      dissociateWithHandler(Redu.MAcc);
+
+    reduCGFunc<KernelName>(KernelFunc, Range, Redu);
+    auto QueueCopy = MQueue;
+    MLastEvent = this->finalize();
+
+    // 2. Run the additional aux kernel as many times as needed to reduce
+    // all partial sums into one scalar.
+    size_t WGSize = Range.get_local_range().size();
+    size_t NWorkItems = NWorkGroups;
+    size_t KernelRun = 1;
+    while (NWorkItems > 1) {
+      // Before creating another kernel, add the event from the previous kernel
+      // to queue.
+      addEventToQueue(QueueCopy, MLastEvent);
+
+      // TODO: here the work-group size is not limited by user's needs,
+      // the better strategy here is to make the work-group-size as big
```

Comment:
Why this is the better strategy?

Reply:
Because user may specify nd_range with small work-group because of some good reasons requirements for user's lambda, for example nd_range=(1M, 4), So, the main kernel must obey that and use WGSize=4, but it is obviously more efficient to use bigger WGSize for aux kernels that only reduce elements and do that much faster when WGSize is big (it also will require less calls of aux kernel, when the WGSize is bigger).

I'll update comment.